FirebaseSessions/Sources/Settings.swift

```diff
@@ -0,0 +1,173 @@
+//
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+extension ApplicationInfoProtocol {
+  var synthesizedVersion: String { return "\(appDisplayVersion) (\(appBuildVersion))" }
+}
+
+private extension Data {
+  var dictionaryValue: [String: AnyObject]? {
+    do {
+      let json = try JSONSerialization.jsonObject(with: self)
+      if let dictionary = json as? [String: AnyObject] {
+        return dictionary
+      } else {
+        Logger
+          .logError("[Settings] Could not cast JSON object as a Dictionary<String, Any>")
+      }
+    } catch {
+      Logger.logError("[Settings] Error: \(error)")
+    }
+    return nil
+  }
+
+  var cacheKeyValue: Settings.CacheKey? {
+    do {
+      let cacheKey = try JSONDecoder().decode(Settings.CacheKey.self, from: self)
+      return cacheKey
+    } catch {
+      Logger.logError("[Settings] Error: \(error)")
+    }
+    return nil
+  }
+}
+
+extension NSLocking {
+  func synchronized<T>(_ closure: () throws -> T) rethrows -> T {
+    lock()
+    defer { unlock() }
+    return try closure()
+  }
+}
+
+protocol SettingsProtocol {
+  func loadCache(googleAppID: String, currentTime: Date)
+  var sessionsEnabled: Bool { get }
+  var samplingRate: Double { get }
+  var sessionTimeout: TimeInterval { get }
+}
+
+class Settings: SettingsProtocol {
+  private static let cacheDurationSecondsDefault: TimeInterval = 60 * 60
+  private let fileManager: SettingsFileManagerProtocol
+  private let appInfo: ApplicationInfoProtocol
+  // Underscored variables may be accessed in different threads,
+  private let lock = NSLock()
+  private var _settingsDictionary: [String: AnyObject]?
+  private var _isCacheKeyExpired: Bool
+  struct CacheKey: Codable {
+    var createdAt: Date
+    var googleAppID: String
+    var appVersion: String
+  }
+
+  var settingsDictionary: [String: AnyObject]? {
+    lock.synchronized {
+      _settingsDictionary
+    }
+  }
+
+  var sessionsEnabled: Bool {
+    guard let enabled = settingsDictionary?["sessions_enabled"]?.boolValue else {
```

Comment:
Lets define the Settings Flag Keys as Constants. Having them inline is hard to parse.